metricbeat/beater/module.go

```diff
@@ -0,0 +1,204 @@
+package beater
+
+import (
+	"expvar"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/filter"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/publisher"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/joeshaw/multierror"
+	"github.com/pkg/errors"
+)
+
+// Expvar metric names.
+const (
+	successesKey = "success"
+	failuresKey  = "failures"
+	eventsKey    = "events"
+)
+
+var (
+	debugf  = logp.MakeDebug("metricbeat")
+	fetches = expvar.NewMap("fetches")
+)
+
+// moduleWrapper contains the Module and the private data associated with
+// running the Module and its MetricSets. The moduleWrapper contains a list
+// of pointer to its associated MetricSets.
+type moduleWrapper struct {
+	mb.Module
+	filters   *filter.FilterList
+	pubClient publisher.Client
+	children  []*metricSetWrapper
+}
+
+// metricSetWrapper contains the MetricSet and the private data associated with
+// running the MetricSet. It contains a pointer to the parent Module.
+type metricSetWrapper struct {
+	mb.MetricSet
+	parent *moduleWrapper // Parent Module.
+	stats  *expvar.Map    // expvar stats for this MetricSet.
```

Comment:
I think we shouldn't document them until we are sure that they will stay. After we implement our own internal metric reporting we can answer this question.